src/plugins/discover/public/application/main/services/discover_saved_search_container.ts

```diff
@@ -318,7 +318,9 @@ function getSearchSourceFieldValueForComparison(
   searchSourceFieldName: keyof SearchSourceFields
 ) {
   if (searchSourceFieldName === 'index') {
-    return searchSource.getField('index')?.id;
+    const query = searchSource.getField('query');
+    // ad-hoc data view id can change, so we rather compare the ES|QL query itself here
```

Comment:
this makes sense, I've got one question, when does the id change in ES|QL and the query.esql doesn't? A quick thought, maybe, when initializing the ad-hoc data view, we could use the used index-pattern (with a prefix) as id. This could also maybe fix the problem of having multiple data views with the same index pattern on a dashboard. So ids wouldn't be random but derived from the used index pattern

Reply:
@kertal I have not found yet why/when it changes

Reply:
maybe if you change the index pattern in the ES|QL query, and then go back to the previous query. Then it will high likely get a new ID

https://github.com/elastic/kibana/blob/5d6812966aac7dd9893a1393112ba027dfa4f32e/src/plugins/discover/public/application/main/utils/get_data_view_by_text_based_query_lang.ts#L35-L37

So if we would assign an id derived by the used index pattern here, then it should be stable (it's not something for this PR, but I'm now continuing thinking, because this might also fix the problem of having multiple data views on a dashboard with the same index pattern) ... FYI @stratoula ... is there a reason not to use a data view id generated by the index pattern of the ES|QL query? like `esql-{indexpattern}` or `esql-{indexpatternHash}`

Reply:
This can actually work Matthias, very good idea 👏

Reply:
I can give it a try tomorrow to see how it will work!